File: answers/leetcode/Sort List.cpp

class Solution {
public:
    ListNode *sortList(ListNode *head) {
		if (NULL == head || NULL == head->next)
		{
			return head;
		}
		ListNode *tail = head, *half = head;
		while (NULL != tail && NULL != tail->next)
		{
			tail = tail->next;
			if (NULL == tail || NULL == tail->next)
			{
				break;
			}
			tail = tail->next;
			half = half->next;
		}
		ListNode *head0 = sortList(half->next);
		half->next = NULL;
		ListNode *head1 = sortList(head);
		ListNode *dull(new ListNode(0)), *current = dull, *current0 = head0, *current1 = head1;
		while (current != NULL && (current0 != NULL || current1 != NULL))
		{
			if (NULL != current0)
			{
				current->next = current0;
			}
			else
			{
				current->next = current1;
			}
			if (NULL != current1 && current1->val <= current->next->val)
			{
				current->next = current1;
				current1 = current1->next;
			} else
			{
				current0 = current0->next;
			}
			current = current->next;
		}
		return dull->next;
    }
};

File: answers/leetcode/Sort List/Sort List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        ListNode left = head;
        ListNode right = head;
        while (right != null && right.next != null && right.next.next != null) {
            left = left.next;
            right = right.next;
            if (right != null) {
                right = right.next;
            }
        }
        right = left.next;
        left.next = null;
        right = this.sortList(right);
        left = this.sortList(head);
        
        ListNode dummy = new ListNode(-1);
        ListNode current = dummy;
        while (left != null || right != null) {
            if (right == null || (left != null && right != null && left.val < right.val)) {
                current.next = left;
                left = left.next;
            } else {
                current.next = right;
                right = right.next;
            }
            current = current.next;
        }
        return dummy.next;
    }
}